apax/model/gmnn.py

```diff
@@ -65,27 +101,44 @@ def __call__(
         offsets,
         perturbation=None,
     ):
-        atomic_energies = self.atomistic_model(R, Z, neighbor, box, offsets, perturbation)
+        # Distances
+        idx = canonicalize_neighbors(neighbor)
+        idx_i, idx_j = idx[0], idx[1]
+
+        # R shape n_atoms x 3
+        R = R.astype(jnp.float64)
+        Ri = R[idx_i]
+        Rj = R[idx_j]
+
+        # dr_vec shape: neighbors x 3
+        if np.all(self.init_box < 1e-6):
+            # reverse conventnion to match TF
+            # distance vector for gas phase training and predicting
+            dr_vec = self.displacement(Rj, Ri)
+        else:
+            # distance vector for training on periodic systems
+            dr_vec = self.displacement(Rj, Ri, perturbation, box)
+            dr_vec += offsets
+
+        # Model Core
+        atomic_energies = self.atomistic_model(dr_vec, Z, idx)
         total_energy = fp64_sum(atomic_energies)
 
         # Corrections
         for correction in self.corrections:
-            energy_correction = correction(R, Z, neighbor, box, offsets, perturbation)
+            energy_correction = correction(dr_vec, Z, idx)
             total_energy = total_energy + energy_correction
+
+        # TODO think of nice abstraction for predicting additional properties
         return total_energy
 
 
 class EnergyDerivativeModel(nn.Module):
-    atomistic_model: AtomisticModel = AtomisticModel()
+    # Alternatively, should this be a function transformation?
+    energy_model: EnergyModel = EnergyModel()
```

Comment:
I'm not too familiar with how `nn.Module` are designed to be subclassed, but creating an instance on class level often can lead to issues, therefore I'd suggest moving this to the `__init__` and using `self.energy_model = EnergyModel()` on an instance bases. But I do remember that some of these didn't like an subclassed `__init__` so maybe don't?

Reply:
yeah there is no overwriting `__init__` for flax and `setup` doesn't take arguments. So as far as I am aware, this is the way to go with flax.